File: python/src/immukv/json_helpers.py
"""Public JSON type definitions for ImmuKV.

Only exports types that users need for custom value parsing.
Internal helper functions are in immukv._internal.json_helpers.
"""

from typing import Callable, Dict, List, TypeVar, Union

V = TypeVar("V")

# Represents any valid JSON value
# This is public API - users need it for type annotations
JSONValue = Union[
    None,
    bool,
    int,
    float,
    str,
    List["JSONValue"],
    Dict[str, "JSONValue"],
]

# Decoder that transforms JSONValue into user's V type
# Users provide this to parse JSON from S3 into their custom types
ValueDecoder = Callable[[JSONValue], V]

# Encoder that transforms user's V type into JSONValue
# Users provide this to serialize their custom types to JSON for S3
ValueEncoder = Callable[[V], JSONValue]

File: typescript/src/jsonHelpers.ts
/**
 * Public JSON type definitions for ImmuKV.
 *
 * Only exports types that users need for custom value parsing.
 * Internal helper functions are in ./internal/jsonHelpers.
 */

/**
 * Represents any valid JSON value.
 */
export type JSONValue =
  | null
  | boolean
  | number
  | string
  | JSONValue[]
  | { [key: string]: JSONValue };

/**
 * Parser that transforms JSONValue into user's V type.
 *
 * @param jsonValue - The JSON value to parse
 * @returns The parsed value of type V
 * @throws Error if the JSON structure is invalid for type V
 */
export type ValueParser<V> = (jsonValue: JSONValue) => V;
